File: java/logbrew-java/src/main/java/co/logbrew/sdk/SpanAttributes.java
package co.logbrew.sdk;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Public payload fields for a span event.
 */
public final class SpanAttributes {
    private final String name;
    private final String traceId;
    private final String spanId;
    private final String status;
    private String parentSpanId;
    private Double durationMs;
    private Map<String, ?> metadata;
    private List<SpanEventSummary> events;
    private List<SpanLinkSummary> links;

    private SpanAttributes(String name, String traceId, String spanId, String status) {
        this.name = name;
        this.traceId = traceId;
        this.spanId = spanId;
        this.status = status;
    }

    /**
     * Creates span attributes with required identity and status fields.
     */
    public static SpanAttributes create(String name, String traceId, String spanId, String status) {
        return new SpanAttributes(name, traceId, spanId, status);
    }

    /**
     * Sets the optional parent span ID.
     */
    public SpanAttributes parentSpanId(String parentSpanId) {
        this.parentSpanId = parentSpanId;
        return this;
    }

    /**
     * Sets the optional span duration in milliseconds.
     */
    public SpanAttributes durationMs(double durationMs) {
        this.durationMs = Double.valueOf(durationMs);
        return this;
    }

    /**
     * Sets optional public metadata values.
     */
    public SpanAttributes metadata(Map<String, ?> metadata) {
        this.metadata = Validation.copyMetadata(metadata);
        return this;
    }

    /**
     * Adds one optional privacy-bounded span event summary.
     */
    public SpanAttributes event(SpanEventSummary event) {
        if (event == null) {
            throw new SdkException("validation_error", "span event summary must be provided");
        }
        if (events == null) {
            events = new ArrayList<>();
        }
        events.add(event);
        SpanEventSummary.requireEventLimit(events.size());
        return this;
    }

    /**
     * Sets optional privacy-bounded span event summaries.
     */
    public SpanAttributes events(Iterable<SpanEventSummary> summaries) {
        if (summaries == null) {
            throw new SdkException("validation_error", "span events must be provided");
        }
        List<SpanEventSummary> copied = new ArrayList<>();
        for (SpanEventSummary summary : summaries) {
            if (summary == null) {
                throw new SdkException("validation_error", "span event summary must be provided");
            }
            copied.add(summary);
            SpanEventSummary.requireEventLimit(copied.size());
        }
        this.events = copied;
        return this;
    }

    /**
     * Adds one optional privacy-bounded span link summary.
     */
    public SpanAttributes link(SpanLinkSummary link) {
        if (link == null) {
            throw new SdkException("validation_error", "span link summary must be provided");
        }
        if (links == null) {
            links = new ArrayList<>();
        }
        links.add(link);
        SpanLinkSummary.requireLinkLimit(links.size());
        return this;
    }

    /**
     * Sets optional privacy-bounded span link summaries.
     */
    public SpanAttributes links(Iterable<SpanLinkSummary> summaries) {
        if (summaries == null) {
            throw new SdkException("validation_error", "span links must be provided");
        }
        List<SpanLinkSummary> copied = new ArrayList<>();
        for (SpanLinkSummary summary : summaries) {
            if (summary == null) {
                throw new SdkException("validation_error", "span link summary must be provided");
            }
            copied.add(summary);
            SpanLinkSummary.requireLinkLimit(copied.size());
        }
        this.links = copied;
        return this;
    }

    Map<String, Object> toMap() {
        Validation.requireNonEmpty("span name", name);
        Validation.requireNonEmpty("span traceId", traceId);
        Validation.requireNonEmpty("span spanId", spanId);
        Validation.requireAllowedValue("span status", status, LogBrewClient.SPAN_STATUSES);
        if (parentSpanId != null) {
            Validation.requireNonEmpty("span parentSpanId", parentSpanId);
        }
        Validation.requireNonNegativeNumber("span durationMs", durationMs);
        Map<String, Object> value = new LinkedHashMap<>();
        value.put("name", name);
        value.put("traceId", traceId);
        value.put("spanId", spanId);
        Validation.putOptionalString(value, "parentSpanId", parentSpanId);
        value.put("status", status);
        if (durationMs != null) {
            value.put("durationMs", durationMs);
        }
        Validation.putOptionalMetadata(value, metadata);
        if (events != null && !events.isEmpty()) {
            List<Map<String, Object>> mappedEvents = new ArrayList<>();
            for (SpanEventSummary event : events) {
                mappedEvents.add(event.toMap());
            }
            value.put("events", mappedEvents);
        }
        if (links != null && !links.isEmpty()) {
            List<Map<String, Object>> mappedLinks = new ArrayList<>();
            for (SpanLinkSummary link : links) {
                mappedLinks.add(link.toMap());
            }
            value.put("links", mappedLinks);
        }
        return value;
    }
}

File: dotnet/logbrew-dotnet/src/LogBrew/SpanAttributes.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;

namespace LogBrew
{
    public sealed class SpanAttributes
    {
        private List<SpanEventSummary>? events;
        private List<SpanLinkSummary>? links;

        private SpanAttributes(string name, string traceId, string spanId, string status)
        {
            Name = name;
            TraceId = traceId;
            SpanId = spanId;
            Status = status;
        }

        public string Name { get; }

        public string TraceId { get; }

        public string SpanId { get; }

        public string Status { get; }

        public string? ParentSpanId { get; private set; }

        public double? DurationMs { get; private set; }

        public IDictionary<string, object?>? Metadata { get; private set; }

        public IReadOnlyList<SpanEventSummary>? Events
        {
            get { return events?.AsReadOnly(); }
        }

        public IReadOnlyList<SpanLinkSummary>? Links
        {
            get { return links?.AsReadOnly(); }
        }

        public static SpanAttributes Create(string name, string traceId, string spanId, string status)
        {
            return new SpanAttributes(name, traceId, spanId, status);
        }

        public SpanAttributes WithParentSpanId(string parentSpanId)
        {
            ParentSpanId = parentSpanId;
            return this;
        }

        public SpanAttributes WithDurationMs(double durationMs)
        {
            DurationMs = durationMs;
            return this;
        }

        public SpanAttributes WithMetadata(IDictionary<string, object?> metadata)
        {
            Metadata = metadata;
            return this;
        }

        public SpanAttributes WithEvents(IEnumerable<SpanEventSummary> summaries)
        {
            ExceptionContract.ThrowIfNull(summaries, nameof(summaries));

            var copied = new List<SpanEventSummary>();
            foreach (var summary in summaries)
            {
                if (summary == null)
                {
                    throw new SdkException("validation_error", "span event summary must be provided");
                }

                copied.Add(summary);
                if (copied.Count > SpanEventSummary.MaxEvents)
                {
                    throw new SdkException("validation_error", "span event summaries must contain at most " + SpanEventSummary.MaxEvents.ToString(CultureInfo.InvariantCulture) + " entries");
                }
            }

            events = copied;
            return this;
        }

        public SpanAttributes WithLinks(IEnumerable<SpanLinkSummary> summaries)
        {
            ExceptionContract.ThrowIfNull(summaries, nameof(summaries));

            var copied = new List<SpanLinkSummary>();
            foreach (var summary in summaries)
            {
                if (summary == null)
                {
                    throw new SdkException("validation_error", "span link summary must be provided");
                }

                copied.Add(summary);
                if (copied.Count > SpanLinkSummary.MaxLinks)
                {
                    throw new SdkException("validation_error", "span link summaries must contain at most " + SpanLinkSummary.MaxLinks.ToString(CultureInfo.InvariantCulture) + " entries");
                }
            }

            links = copied;
            return this;
        }

        public SpanAttributes WithLink(SpanLinkSummary summary)
        {
            ExceptionContract.ThrowIfNull(summary, nameof(summary));

            var copied = links == null ? new List<SpanLinkSummary>() : new List<SpanLinkSummary>(links);
            copied.Add(summary);
            if (copied.Count > SpanLinkSummary.MaxLinks)
            {
                throw new SdkException("validation_error", "span link summaries must contain at most " + SpanLinkSummary.MaxLinks.ToString(CultureInfo.InvariantCulture) + " entries");
            }

            links = copied;
            return this;
        }

        public SpanAttributes WithEvent(SpanEventSummary summary)
        {
            ExceptionContract.ThrowIfNull(summary, nameof(summary));

            var copied = events == null ? new List<SpanEventSummary>() : new List<SpanEventSummary>(events);
            copied.Add(summary);
            if (copied.Count > SpanEventSummary.MaxEvents)
            {
                throw new SdkException("validation_error", "span event summaries must contain at most " + SpanEventSummary.MaxEvents.ToString(CultureInfo.InvariantCulture) + " entries");
            }

            events = copied;
            return this;
        }

        internal OrderedJsonObject ToJsonObject()
        {
            Validation.RequireNonEmpty("span name", Name);
            Validation.RequireNonEmpty("span traceId", TraceId);
            Validation.RequireNonEmpty("span spanId", SpanId);
            Validation.RequireAllowedValue("span status", Status, LogBrewClient.SpanStatuses);
            if (ParentSpanId != null)
            {
                Validation.RequireNonEmpty("span parentSpanId", ParentSpanId);
            }

            if (DurationMs.HasValue && (DurationMs.Value < 0 || double.IsNaN(DurationMs.Value) || double.IsInfinity(DurationMs.Value)))
            {
                throw new SdkException("validation_error", "span durationMs must be non-negative");
            }

            var payload = new OrderedJsonObject()
                .Add("name", Name)
                .Add("traceId", TraceId)
                .Add("spanId", SpanId);
            payload.AddIfNotNull("parentSpanId", ParentSpanId);
            payload.Add("status", Status);
            if (DurationMs.HasValue)
            {
                payload.Add("durationMs", DurationMs.Value);
            }

            payload.AddMetadata(Metadata);
            if (events != null && events.Count > 0)
            {
                payload.Add("events", events.Select(summary => summary.ToJsonObject()).ToList());
            }

            if (links != null && links.Count > 0)
            {
                payload.Add("links", links.Select(summary => summary.ToJsonObject()).ToList());
            }

            return payload;
        }
    }

    public sealed class SpanLinkSummary
    {
        internal const int MaxLinks = 8;

        private SpanLinkSummary(string traceId, string spanId, string traceFlags)
        {
            TraceId = traceId;
            SpanId = spanId;
            TraceFlags = traceFlags;
            Sampled = (Convert.ToInt32(traceFlags, 16) & 1) == 1;
        }

        public string TraceId { get; }

        public string SpanId { get; }

        public string TraceFlags { get; }

        public bool Sampled { get; }

        public IDictionary<string, object?>? Metadata { get; private set; }

        public static SpanLinkSummary FromTraceparent(string traceparent)
        {
            var context = Traceparent.Parse(traceparent);
            return new SpanLinkSummary(context.TraceId, context.ParentSpanId, context.TraceFlags);
        }

        public static SpanLinkSummary Create(string traceId, string spanId, string traceFlags)
        {
            var normalizedTraceparent = Traceparent.Parse(Traceparent.Create(traceId, spanId, traceFlags));
            return new SpanLinkSummary(normalizedTraceparent.TraceId, normalizedTraceparent.ParentSpanId, normalizedTraceparent.TraceFlags);
        }

        public SpanLinkSummary WithMetadata(IDictionary<string, object?> metadata)
        {
            Metadata = SpanSummaryMetadata.CopyMetadata(metadata, "span link");
            return this;
        }

        internal OrderedJsonObject ToJsonObject()
        {
            var payload = new OrderedJsonObject()
                .Add("traceId", TraceId)
                .Add("spanId", SpanId)
                .Add("sampled", Sampled);
            payload.AddMetadata(Metadata);
            return payload;
        }

        internal SpanLinkSummary WithSafeMetadata(IDictionary<string, object?> metadata)
        {
            Metadata = metadata;
            return this;
        }
    }

    public sealed class SpanEventSummary
    {
        internal const int MaxEvents = 8;

        private SpanEventSummary(string name)
        {
            Name = name;
        }

        public string Name { get; }

        public string? Timestamp { get; private set; }

        public IDictionary<string, object?>? Metadata { get; private set; }

        public static SpanEventSummary Create(string name)
        {
            return new SpanEventSummary(name);
        }

        public SpanEventSummary WithTimestamp(string timestamp)
        {
            Timestamp = timestamp;
            return this;
        }

        public SpanEventSummary WithMetadata(IDictionary<string, object?> metadata)
        {
            Metadata = CopyMetadata(metadata);
            return this;
        }

        internal OrderedJsonObject ToJsonObject()
        {
            Validation.RequireNonEmpty("span event name", Name);
            if (Timestamp != null)
            {
                Validation.RequireTimestamp(Timestamp);
            }

            var payload = new OrderedJsonObject().Add("name", Name);
            payload.AddIfNotNull("timestamp", Timestamp);
            payload.AddMetadata(Metadata);
            return payload;
        }

        private static Dictionary<string, object?> CopyMetadata(IDictionary<string, object?> metadata)
        {
            return SpanSummaryMetadata.CopyMetadata(metadata, "span event");
        }
    }

    internal static class SpanSummaryMetadata
    {
        internal static Dictionary<string, object?> CopyMetadata(IDictionary<string, object?> metadata, string label)
        {
            ExceptionContract.ThrowIfNull(metadata, nameof(metadata));

            var copied = new Dictionary<string, object?>(StringComparer.Ordinal);
            foreach (var item in metadata)
            {
                Validation.RequireNonEmpty(label + " metadata key", item.Key);
                if (!Validation.IsMetadataValue(item.Value))
                {
                    throw new SdkException("validation_error", label + " metadata value for " + item.Key + " must be a string, number, boolean, or null");
                }

                copied[item.Key] = item.Value;
            }

            return copied;
        }
    }
}
